File: python/producer.py
from kafka import SimpleProducer, KafkaClient
import avro.schema
import io, random
from avro.io import DatumWriter
import cProfile

# Kafka settings
kafkaHost = "localhost:9092"
kafkaTopic = "avro"

kafka = KafkaClient(kafkaHost)
producer = SimpleProducer(kafka)

schema_path="../schemas/user.avsc"
schema = avro.schema.parse(open(schema_path).read())

writer = avro.io.DatumWriter(schema)
buffer = io.BytesIO()
encoder = avro.io.BinaryEncoder(buffer)

datum = {"created_at": "2015-10-21T09:47:50-04:00", "name": "123", "favorite_color": "111", "favorite_number": random.randint(0,10)}
writer.write(datum, encoder)

raw_bytes = buffer.getvalue()

producer.send_messages(kafkaTopic, raw_bytes)

File: ruby/producer.rb
require "avro"
require "kafka"

KAFKA_HOST = "localhost:9092"
KAFKA_TOPIC = "avro"

kafka = Kafka.new(seed_brokers: [KAFKA_HOST])
schema = Avro::Schema.parse(File.open("../schemas/user.avsc", "rb").read)

def encode(schema)
  writer = Avro::IO::DatumWriter.new(schema)
  buffer = StringIO.new("".force_encoding("BINARY"))
  encoder = Avro::IO::BinaryEncoder.new(buffer)

  # use old hash style, won"t work otherwise
  datum = { "created_at" => "2015-10-21T09:47:50-04:00", "name" => "John", "favorite_color" => "red", "favorite_number" => 1 }

  writer.write(datum, encoder)
  buffer.string
end

kafka.deliver_message(encode(schema), topic: KAFKA_TOPIC)
